java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java

```diff
@@ -0,0 +1,1464 @@
+   * elements in the vector. This method must be called prior to using the ValueVector.
+   *
+   * @param totalBytes desired total memory capacity
+   * @param valueCount the desired number of elements in the vector
+   * @throws org.apache.arrow.memory.OutOfMemoryException if memory allocation fails
+   */
+  @Override
+  public void allocateNew(long totalBytes, int valueCount) {
+    assert totalBytes >= 0;
+
+    checkDataBufferSize(totalBytes);
+
+    /* we are doing a new allocation -- release the current buffers */
+    clear();
+
+    try {
+      allocateBytes(totalBytes, valueCount);
+    } catch (Exception e) {
+      clear();
+      throw e;
+    }
+  }
+
+  @Override
+  public void allocateNew(int valueCount) {
+    allocateNew(lastValueAllocationSizeInBytes, valueCount);
+  }
+
+  /* Check if the data buffer size is within bounds. */
+  private void checkDataBufferSize(long size) {
+    if (size > MAX_BUFFER_SIZE || size < 0) {
+      throw new OversizedAllocationException("Memory required for vector " +
+          "is (" + size + "), which is overflow or more than max allowed (" + MAX_BUFFER_SIZE + "). " +
+          "You could consider using LargeVarCharVector/LargeVarBinaryVector for large strings/large bytes types");
+    }
+  }
+
+  /* allocate the inner buffers */
+  private void allocateBytes(final long valueBufferSize, final int valueCount) {
+    /* allocate data buffer */
+    viewBuffer = allocator.buffer(valueBufferSize);
+    viewBuffer.readerIndex(0);
+
+    validityBuffer = allocator.buffer((valueCount + 7) / 8);
+    initValidityBuffer();
+
+    lastValueCapacity = getValueCapacity();
+    lastValueAllocationSizeInBytes = capAtMaxInt(viewBuffer.capacity());
+  }
+
+  /**
+   * Resize the vector to increase the capacity. The internal behavior is to
+   * double the current value capacity.
+   */
+  @Override
+  public void reAlloc() {
+    reallocViewBuffer();
+    reallocViewReferenceBuffer();
+    reallocValidityBuffer();
+  }
+
+  /**
+   * Reallocate the data buffer. Data Buffer stores the actual data for
+   * VIEWVARCHAR or VIEWVARBINARY elements in the vector. The behavior is to double
+   * the size of buffer.
+   * @throws OversizedAllocationException if the desired new size is more than
+   *                                      max allowed
+   * @throws OutOfMemoryException if the internal memory allocation fails
+   */
+  public void reallocViewBuffer() {
+    long currentViewBufferCapacity = viewBuffer.capacity();
+
+    long newAllocationSize = currentViewBufferCapacity * 2;
+    if (newAllocationSize == 0) {
+      if (lastValueAllocationSizeInBytes > 0) {
+        newAllocationSize = lastValueAllocationSizeInBytes;
+      } else {
+        newAllocationSize = INITIAL_BYTE_COUNT * 2L;
+      }
+    }
+
+    reallocViewBuffer(newAllocationSize);
+  }
+
+  /**
+   * Reallocate the data buffer for reference buffer.
+   */
+  public void reallocViewReferenceBuffer() {
```

Comment:
I think this would be my bad naming. What I meant was the external buffer which keeps the long strings. I used the reference, as the viewBuffer references this via offset and index.

Reply:
The spec calls this the data buffer. Let's stick to the names there.

Reply:
Sure @lidavidm 
I saw the term data buffer in a util where we use a single buffer, split it and make validity and offset buffers. Just curious, would that need a renaming? As it was a bit confusing. Not related to this in anyway, but just for understanding.